x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/pivot/DateHistogramGroupSource.java

```diff
@@ -248,6 +256,71 @@ public static DateHistogramGroupSource fromXContent(final XContentParser parser,
+        String expression = interval.toString();
+        String epochFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli());
+        String epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + interval.estimateMillis());
+        if (DateHistogramAggregationBuilder.DATE_FIELD_UNITS.containsKey(expression)) {
+            // The estimated rounding defaults to 30 days for monthly, EPOCH starts in January, which has 31 days.
+            if (Rounding.DateTimeUnit.MONTH_OF_YEAR.equals(DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(expression))) {
+                epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + THIRTY_ONE_DAYS_MS);
+            }
+        } else {
+            // This should never really happen in practice. date_histogram does not support multiple month intervals
+            if (expression.endsWith("M")) {
+                throw new IllegalArgumentException("only single month intervals are supported");
+            }
+            final String normalized = expression.toLowerCase(Locale.ROOT).trim();
+            long millisecondLength;
+            if (normalized.endsWith("ms")) {
+                millisecondLength = unitMilliseconds(normalized, "ms", TimeUnit.MILLISECONDS).getMillis();
+            } else if (normalized.endsWith("s")) {
+                millisecondLength = unitMilliseconds(normalized, "s", TimeUnit.SECONDS).getMillis();
+            } else if (normalized.endsWith("m")) {
+                millisecondLength = unitMilliseconds(normalized, "m", TimeUnit.MINUTES).getMillis();
+            } else if (normalized.endsWith("h")) {
+                millisecondLength = unitMilliseconds(normalized, "h", TimeUnit.HOURS).getMillis();
+            } else if (normalized.endsWith("d")) {
+                millisecondLength = unitMilliseconds(normalized, "d", TimeUnit.DAYS).getMillis();
+            } else {
+                return true;
+            }
+            epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + millisecondLength);
+        }
+        return epochFormat.equals(epochPlusIntervalFormat);
+    }
+
+    // We round down and + 1 for all format types to force the interval to make a change in the format time unit value
+    private static TimeValue unitMilliseconds(String value, String suffix, TimeUnit unit) {
+        long valueAsLong = parse(value, suffix);
+        switch (unit) {
+            case MILLISECONDS:
+                return new TimeValue((valueAsLong % 999) + 1, unit);
+            case SECONDS:
+            case MINUTES:
+                return new TimeValue((valueAsLong % 59) + 1, unit);
+            case HOURS:
+                return new TimeValue((valueAsLong % 23) + 1, unit);
+            case DAYS:
+                return new TimeValue((valueAsLong % 30) + 1, unit);
+            default:
+                throw new IllegalArgumentException("failed to determine TimeUnit [" + unit + "]");
```

Comment:
This error message isn't really true.  `TimeUnit` has two other values, namely `MICROSECONDS` and `NANOSECONDS`.  This might be better as:

```
    assert false : "unsupported TimeUnit [" + unit + "]";
    return TimeValue.ZERO;
```

(But note in another comment I've suggested completely removing this method, so don't worry about this comment unless my main idea doesn't work for some reason.)